pkg/webhook/validator.go

```diff
@@ -0,0 +1,74 @@
+package webhook
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	liberr "github.com/konveyor/controller/pkg/error"
+	"github.com/konveyor/controller/pkg/logging"
+	api "github.com/konveyor/forklift-controller/pkg/apis/forklift/v1beta1"
+	"github.com/konveyor/forklift-controller/pkg/controller/provider/container"
+	core "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+const (
+	// Name.
+	Name = "admission-webhook"
+)
+
+// Package logger.
+var log = logging.WithName(Name)
+
+type Handler struct {
+	Client  client.Client
+	decoder *admission.Decoder
+}
+
+//
+// Build custom provider with parameters form secret
+func buildProvider(provider *api.Provider, providerType *api.ProviderType, secret *core.Secret) {
+	provider.Spec.URL = string(secret.Data["url"])
+	provider.Spec.Type = providerType
+	provider.Name = secret.Name
+	provider.Namespace = secret.Namespace
+}
+
+//
+// Admission webhook handler
+// Tests the connection to the provider before creating the secret
+func (a *Handler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	secret := &core.Secret{}
+	err := a.decoder.Decode(req, secret)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	// Get the provider type on which start the tests
+	createdForProviderType, ok := secret.GetLabels()["createdForProviderType"]
+	if !ok {
+		return admission.Errored(http.StatusBadRequest, liberr.New("The label 'createdForProviderType' is not set on secret"))
```

Comment:
this should be dropped - we still want to accept secrets without this label, and validate those that include the label
I played with this change locally and besides that it works well - I'll include this in #457  along with the 'forklift-api' pod

Reply:
This might be a problem because the secrets don't specify for which provider type they are. We can choose by the data fields. For example, only rhv has cacert or vmware has thumbprint but this might not be 100% reliable. Or reading the url parsing to the provider type according to it.

Reply:
I'm ok with checking the provider type the way you do it now, if the label exists
but the api server can get other secrets, unrelated to MTV providers, that shouldn't be rejected

Reply:
Those will not trigger the webhook please see [1]. Only secrets with the label `createdForResourceType: providers` will trigger the webhook.
[1] https://github.com/konveyor/forklift-operator/pull/247/files#diff-0ae11594eeec8b4ee7eebf3797d21139a635edf688a55164f674355f49c47ff3R13-R16

Reply:
ack, you see - that's what I don't like in this repo-per-component structure - that's it's easy to miss such things.. another argument for having an operator that is closer to the "backend" components..